### ISSUE TITLE
run_swat2012() on Linux failed to find function "%//%"

### ISSUE THREAD

**Issue by Ericoncy** [chrisschuerz/SWATrunR#70]
Hi Chris,

I am trying to transfer the simulation from my laptop to the HPC (Linux clusters). The project worked well on my device, but had an error on HPC (I used the Linux version of SWAT executable file):
# ================================================================================
Thread 1 of 1   Time elapsed: 2S   Time remaining: 0S    
                                                                            
 Completed 1 thread in 2S 
Performing 1 simulation on 1 cores: 

 Simulation 1 of 1   Time elapsed: 1S   Time remaining: 0S    Error in { : task 1 failed - "could not find function "%//%""
Calls: run_swat2012 -> %dopar% -> <Anonymous>
Execution halted
# ================================================================================
I checked the source code and it seems the foreach loop from swat2012_run_swat.R at Line 226 cannot recognize the pre-defined %//% function generated in general_functions.R. BUT this is not an issue on Windows.
Could you give some insight to fix this issue on Linux? I am not sure which step I missed to configure the model on Linux.

Thanks!
Ziming

**Comment by Ericoncy** [comment]
Hi all,

Just in case someone also interesting with that.
Previous I just simply assign the library path in R scripts through .libPaths("User-Defined Path"), BUT it seems cannot set the library path on each worker.
Another way is directly modify the .bash_profile, adding a line: export R_LIBS_USER="Personal library path here".

Thanks!

**Comment by banomehdi** [comment]
Vielen Dank für Ihre Nachricht. Ich bin derzeit auf Urlaub und am 19.02.24 wieder erreichbar.

Thank you for your message. I am currently out of the office and will be back on 19 Feb. During this time I will only check my emails irregularly,

Herzliche Grüße / Best regards,
Bano Mehdi-Schulz

**Comment by chrisschuerz** [comment]
@Ericoncy thanks for showing a workaround. I have not used `SWATrunR` for a while on Linux, but not finding functions which are defined in the R package was never an issue before. The same error was however reported to me recently in an email. 
I plan to replace these infix functions (e.g. %//%) by simple paste functions in the code, but not sure when I have time for that.
I hope that this will solve this issue and other functions won't pop up then instead which cannot be found.

**Comment by Ericoncy** [comment]
Hi Chris,
It runs well now on Linux.
Also a quick question you may have some initial ideas. I read your replies about coupling SWATrunR with other auto-searching optimization algorithm (e.g., GA, PSO). It seems not straightforward if we want to use these algorithm packages in R with parallel computation, unless re-writing them with SWATrunR. Have you coupled SWATrunR with other auto-optimization algorithm before, using parallelization on HPC?
Thanks!

**Comment by chrisschuerz** [comment]
I still think this can be challenging. It will mainly depend on the design of the optimizer. I can think of a "batchwise" parallel workflow that could work with SWATrunR. As far as I can remember @drkupi presented a batchwise parallel optimizer a while ago which could conceptually work with SWATrunR. But I am pretty sure that if he has some code for that it will be in python :)

So the only way where you can make use of the parallelization implemented in SWATrunR is to send batches of parameter combinations to `run_swat*()` and return the performance vector of that batch of parameter combinations to the optimizer.

If the optimizer has a parallel option implemented, then the challenge will be to take care that the optimizer generates parallel copies of the SWAT project. Otherwise you will run into access conflicts in the project folder.

So long story short, no I did not really use optimizers recently. Sorry that I cannot offer more help here.

**Comment by Ericoncy** [comment]
Thanks Chris! That helped a lot.
Have a good day :)
